torch_np/_funcs.py

```diff
@@ -1,21 +1,1043 @@
+    (x,) = _util.typecast_tensors((x,), int_dtype, casting="safe")
+
+    result = torch.bincount(x, weights, minlength)
+    return result
+
+
+@normalizer
+def where(
+    condition: ArrayLike,
+    x: Optional[ArrayLike] = None,
+    y: Optional[ArrayLike] = None,
+    /,
+):
+    selector = (x is None) == (y is None)
+    if not selector:
+        raise ValueError("either both or neither of x and y should be given")
+
+    if condition.dtype != torch.bool:
+        condition = condition.to(torch.bool)
+
+    if x is None and y is None:
+        result = torch.where(condition)
+    else:
+        try:
+            result = torch.where(condition, x, y)
+        except RuntimeError as e:
+            raise ValueError(*e.args)
+    return result
+
+
+###### module-level queries of object properties
+
+
+@normalizer
+def ndim(a: ArrayLike):
+    return a.ndim
+
+
+@normalizer
+def shape(a: ArrayLike):
+    return tuple(a.shape)
+
+
+@normalizer
+def size(a: ArrayLike, axis=None):
+    if axis is None:
+        return a.numel()
+    else:
+        return a.shape[axis]
+
+
+###### shape manipulations and indexing
+
+
+@normalizer
+def expand_dims(a: ArrayLike, axis):
+    shape = _util.expand_shape(a.shape, axis)
+    tensor = a.view(shape)  # never copies
+    return tensor
+
+
+@normalizer
+def flip(m: ArrayLike, axis=None):
+    # XXX: semantic difference: np.flip returns a view, torch.flip copies
+    if axis is None:
+        axis = tuple(range(m.ndim))
+    else:
+        axis = _util.normalize_axis_tuple(axis, m.ndim)
+    return torch.flip(m, axis)
+
+
+@normalizer
+def flipud(m: ArrayLike):
+    result = torch.flipud(m)
+    return result
+
+
+@normalizer
+def fliplr(m: ArrayLike):
+    result = torch.fliplr(m)
+    return result
+
+
+@normalizer
+def rot90(m: ArrayLike, k=1, axes=(0, 1)):
+    axes = _util.normalize_axis_tuple(axes, m.ndim)
+    return torch.rot90(m, k, axes)
+
+
+# ### broadcasting and indices ###
+
+
+@normalizer
+def broadcast_to(array: ArrayLike, shape, subok: SubokLike = False):
+    result = torch.broadcast_to(array, size=shape)
+    return result
+
+
+from torch import broadcast_shapes
+
+
+@normalizer
+def broadcast_arrays(*args: ArrayLike, subok: SubokLike = False):
+    res = torch.broadcast_tensors(*args)
+    return res
+
+
+# FIXME: this needs fixing
+def unravel_index(indices, shape, order="C"):
+    # cf https://github.com/pytorch/pytorch/pull/66687
+    # this version is from
+    # https://discuss.pytorch.org/t/how-to-do-a-unravel-index-in-pytorch-just-like-in-numpy/12987/3
+    if order != "C":
+        raise NotImplementedError
+    result = []
+    for index in indices:
+        out = []
+        for dim in reversed(shape):
+            out.append(index % dim)
+            index = index // dim
+        result.append(tuple(reversed(out)))
+    return result
+
+
+def ravel_multi_index(multi_index, dims, mode="raise", order="C"):
+    return sum(idx * dim for idx, dim in zip(multi_index, dims))
```

Comment:
We use the normalization approach (among other things) to handle the fact that NumPy often accepts anythng thats smells like an array in place of an array (e.g. a list of lists) and then returns an array. This function would deviate from that behaviour because we are not normalizing it. In general, whenever we can, we should use the normalizer to avoid these sly deviations from the NumPy semantics.

Reply:
OK, let's remove it for now and add it back to the TODO priority queue.